packages/rabbitmq/src/TxSubmitWorker.ts

```diff
@@ -0,0 +1,282 @@
+    const exitResolver = this.#exitResolver;
+    this.#exitResolver = undefined;
+
+    try {
+      this.#dependencies.logger!.info(`${moduleName} shutdown: closing RabbitMQ channel`);
+
+      try {
+        if (this.#consumerTag) {
+          const consumerTag = this.#consumerTag;
+          this.#consumerTag = undefined;
+
+          await this.#channel?.cancel(consumerTag);
+        }
+      } catch (error) {
+        this.logError(error);
+      }
+
+      this.#dependencies.logger!.info(`${moduleName} shutdown: closing RabbitMQ connection`);
+
+      try {
+        await this.#connection?.close();
+      } catch (error) {
+        this.logError(error);
+      }
+
+      this.#dependencies.logger!.info(`${moduleName}: shutdown completed`);
+      this.#channel = undefined;
+      this.#connection = undefined;
+      this.#consumerTag = undefined;
+      this.#continueForever = false;
+      this.#status = 'idle';
+    } finally {
+      // Only logging functions could throw an error here...
+      // Although this is almost impossible, we want to be sure exitResolver is called
+      exitResolver?.();
+    }
+  }
+
+  /**
+   * Wrapper to log errors from try/catch blocks
+   *
+   * @param {any} error the error to log
+   */
+  private logError(error: unknown, isAsync = false) {
+    const errorMessage =
+      // eslint-disable-next-line prettier/prettier
+      error instanceof Error ? error.message : (typeof error === 'string' ? error : JSON.stringify(error));
+    const errorObject = { error: error instanceof Error ? error.name : 'Unknown error', isAsync, module: moduleName };
+
+    this.#dependencies.logger!.error(errorObject, errorMessage);
+    if (error instanceof Error) this.#dependencies.logger!.debug(`${moduleName}:`, error.stack);
+  }
+
+  /**
+   * The infinite loop to perform serial tx submission
+   */
+  private async infiniteLoop() {
+    this.#continueForever = true;
+    this.#status = 'connected';
+
+    while (this.#continueForever) {
+      const message = await this.#channel?.get(TX_SUBMISSION_QUEUE);
+
+      // If there is a message, handle it, otherwise wait #pollingCycle ms
+      await (message
+        ? this.submitTx(message)
+        : new Promise((resolve) => setTimeout(resolve, this.#config.pollingCycle!)));
+    }
+  }
+
+  /**
+   * Submit a tx to the provider and ack (or nack) the message
+   *
+   * @param {Message} message the message containing the transaction
+   */
+  private async submitTx(message: Message) {
+    try {
+      const counter = ++this.#counter;
+      const { content } = message;
+
+      this.#dependencies.logger!.info(`${moduleName}: submitting tx`);
+      this.#dependencies.logger!.debug(`${moduleName}: tx ${counter} dump:`, content.toString('hex'));
+      await this.#dependencies.txSubmitProvider.submitTx(new Uint8Array(content));
+
+      this.#dependencies.logger!.debug(`${moduleName}: ACKing RabbitMQ message ${counter}`);
+      this.#channel?.ack(message);
```

Comment:
@rhyslbw how reliable do we need this service to be? Is the mempool preserved if the node dies for some reason?

Should rollbacks be handled in this service or the frontend? :thinking:

We could use ogmios chainsync to:
- await for it to be included in a block + n blocks before acking
- nacking after n blocks of tx not being included

If we generalize such 'confirmation service' as a dependency to TxSubmitWorker then we can simulate various scenarios in tests reliably

Reply:
I had originally scoped this to be a basic form of the worker, with optimisations and reliability improvements to come when introducing https://ogmios.dev/mini-protocols/local-tx-monitor/. [TxMonitor](https://github.com/CardanoSolutions/ogmios/tree/master/clients/TypeScript/packages/client/src/TxMonitor) after SPIKE ADP-1547

So how about we get the current set of commits squashed and merged down, then @iccicci can then take ADP-1822 with next once the SPIKE is complete. c66494a57d73711083cf3381761848f9da3f67e5 is at least one fix commit that would be good to get to _master_.

Reply:
But to answer your questions @mkazlauskas:

> how reliable do we need this service to be?
On-chain confirmation.

> Is the mempool preserved if the node dies for some reason?

The more suitable question for me is: how do we model the scenario of node crashing, regardless of it’s own recovery, so we can deal with packet loss or other unforeseen scenarios.

> Should rollbacks be handled in this service or the frontend? thinking To me it seems like this backend service is a better place to do that.

Agree, for tx submission, chain rollbacks can be opaque. We could consider a more sophisticated process where we optimistically ack by using a configurable depth, push the tx (or just a hash just for auditing purpose) with an expiry based on the estimated TTL or the volatile block range cross-over (whichever comes first). This result is guaranteed submission if there’s a rollback of any length.

> maybe we could use ogmios chainsync (or the db-sync provider) to:
> await for it to be included in a block + n blocks before acking

ChainSync, as that's where we have rollback awareness, and it’s more direct to what we want to achieve with fewer moving parts.

> nacking after n blocks of tx not being included

This one is tricky, as it’s quite variable, particularly if the mempool moves to a mode of fee priority. The way I’ve specified below allows us to use as few resources and delays as possible.

> If we generalize such 'confirmation service' as a dependency to TxSubmitWorker then we can simulate various scenarios in tests reliably

Great idea! `TxSubmitConfirmer`?